firedrake/slate/slac/utils.py

```diff
@@ -136,6 +151,213 @@ def visit_Symbol(self, o, *args, **kwargs):
+        A, = tensor.operands
+        _A = node_dict[A]
+        self.builder.create_index(A.shape, str(A)+"newtrans")
+        new_indices = self.builder.gem_indices[str(A)+"newtrans"]
+        var_A = self.index_tensor(_A, new_indices)
+        ret = ComponentTensor(var_A, new_indices[::-1])
+        return ret
+
+    @slate_to_gem.register(sl.Mul)
+    def slate_to_gem_mul(self, tensor, node_dict):
+        A, B = tensor.operands
+        var_A, var_B = node_dict[A], node_dict[B]  # gem representations
+
+        # New indices are necessary in case as Tensor gets multiplied with itself.
+        self.builder.create_index(A.shape, str(A)+"newmulA"+str(tensor))
+        new_indices_A = self.builder.gem_indices[str(A)+"newmulA"+str(tensor)]
+        self.builder.create_index(B.shape, str(B)+"newmulB"+str(tensor))
+        new_indices_B = self.builder.gem_indices[str(B)+"newmulB"+str(tensor)]
+
+        if len(A.shape) == len(B.shape) and A.shape[1] == B.shape[0]:
+            var_A = self.index_tensor(var_A, new_indices_A)
+            var_B = self.index_tensor(var_B, (new_indices_A[1], new_indices_B[1]))
+
+            prod = Product(var_A, var_B)
+            sum = IndexSum(prod, (new_indices_A[1],))
+            return ComponentTensor(sum, (new_indices_A[0], new_indices_B[1]))
+
+        elif len(A.shape) > len(B.shape) and A.shape[1] == B.shape[0]:
+            var_A = self.index_tensor(var_A, new_indices_A)
+            var_B = self.index_tensor(var_B, (new_indices_A[1],))
+
+            prod = Product(var_A, var_B)
+            sum = IndexSum(prod, (new_indices_A[1],))
+            return ComponentTensor(sum, (new_indices_A[0],))
+        else:
+            return[]
+
+    @slate_to_gem.register(sl.Block)
+    def slate_to_gem_blocks(self, tensor, node_dict):
+
+        A, = tensor.operands
+        node = node_dict[A]
+
+        # get first block while handling ranges handle ranges
+        first_ind = ()
+        for index in tensor._indices:
+            if type(index) != range:
+                first_ind += (index,)
+            else:
+                first_ind += (index.start,)
+        first_block = tuple(range(ten, ten+1) for ten in first_ind)
+        index_offset = ()
+
+        # i points to the dim
+        # idx points to the shape of that block matrix in that dim
+        for i, idx in enumerate(first_block):
+            if idx.start == 0:
+                index_offset += (0, )
+            else:
+                index_offset += ((sum(A.shapes[i][:idx.start])), )
+
+        index_extent = tuple(sum(shape) for shape in tensor.shapes.values())
+        self.builder.create_index(index_extent, tensor)
+        gem_index = self.builder.gem_indices[tensor]
+
+        dim2idxs = []
+        for i, dim in enumerate(tensor.shapes.keys()):
+            dim2idxs += ((index_offset[i], ((gem_index[i], 1), )),)
+
+        return ComponentTensor(FlexiblyIndexed(node, tuple(dim2idxs)), gem_index)
```

Comment:
FWIW, I think this isn't right for cases which the slate frontend language supports. Namely non-contiguous block indices. I think it's permissible to say `B = A.blocks[0:2, (0, 1, 4)]` (say). But `FlexiblyIndexed` can only handle contiguous ranges. I would suggest that this is handled in two modes. 

1. contiguous ranges (just use a view via `FlexibilyIndexed`
2. general indexing, generate a temporary plus some copy-in code that splats the separate views in. Relatedly, I think this code could be simpler (something like the below, if we ensure indices in slate.Block are always ranges or tuples)

```
@_slate2gem.register(slate.Block)
def _slate2gem_block(expr, self):
    child, = map(self, expr.children)

    # Unpack ranges
    indices = tuple(tuple(idx) for idx in expr._indices)
    if any(i != tuple(range(i[0], i[-1] + 1)) for i in indices):
        # Handle general (non range) indexing
        # This needs copy-in code generation from elsewhere.
        shapes = expr.children[0].shapes.values()
        child_indices = [[(sum(extents[:idx]), sum(extents[idx]))
                          for idx in idxs]
                         for idxs, extents in zip(indices, shapes)]
        blocks = [gem.view(child, *(slice(*view) for view in block))
                  for block in itertools.product(*child_indices)]
        name = f"B{len(self.blockcopy)}"
        var = gem.Variable(name, expr.shape)
        self.blockcopy[var] = numpy.asarray(blocks).reshape(*map(len, indices))
        return var
    else:
        # Just a view
        slices = []
        for idxs, extents in zip(indices, expr.children[0].shapes.values()):
            sidx = idxs[0]
            eidx = idxs[-1]
            slices.append(slice(sum(extents[:sidx]), sum(extents[:eidx+1])))
        return gem.view(child, *slices)
```

Reply:
The previous slate compiler has a an exception for non-contiguous arrays, see https://github.com/firedrakeproject/firedrake/blob/418eaea8e307de39983cf9eb60caa8e01d266212/firedrake/slate/slac/compiler.py#L980. But that does not stop us from implementing that now, I guess.

Reply:
Ah, I forgot that. In which case we can carry over the same thing for now (and the else branch is sufficient).